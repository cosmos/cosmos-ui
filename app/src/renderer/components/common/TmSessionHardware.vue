<template lang="pug">
.tm-session: .tm-session-container
  .tm-session-header
    a(@click="setState('welcome')"): i.material-icons arrow_back
    .tm-session-title Sign In
    a(@click="help"): i.material-icons help_outline
  .tm-session-main
    hardware-state(v-if="status == 'connect'" @click.native="setStatus('detect')"
      icon="usb" value="Please plug in your Ledger Wallet")
    hardware-state(v-if="status == 'detect'" @click.native="setStatus('success')"
      icon="rotate_right" spin="true" value="Detecting your Ledger Wallet")
    hardware-state(v-if="status == 'success'" @click.native="onSubmit"
      icon="check_circle" value="Ledger Wallet successfully loaded")
  .tm-session-footer
</template>

<script>
import HardwareState from "common/TmHardwareState"
export default {
  name: `tm-session-hardware`,
  components: { HardwareState },
  data: () => ({ status: `connect` }),
  methods: {
    help() {
      this.$store.commit(`setModalHelp`, true)
    },
    setState(value) {
      this.$store.commit(`setModalSessionState`, value)
    },
    setStatus(value) {
      this.status = value
    },
    onSubmit() {
      this.$store.commit(`setModalSession`, false)
      this.$store.commit(`notify`, {
        title: `Welcome back!`,
        body: `You are now signed in to your Cosmos account.`
      })
      this.$store.dispatch(`signIn`, { password: this.fields.signInPassword })
    }
  }
}
</script>
