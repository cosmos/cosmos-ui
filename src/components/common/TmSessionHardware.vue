<template>
  <div class="session">
    <div class="session-header">
      <a @click="goBack">
        <i class="material-icons session-back">arrow_back</i>
      </a>
      <a @click="close">
        <i class="material-icons session-close">close</i>
      </a>
    </div>
    <h2 class="session-title">
      Use my Ledger Nano
    </h2>

    <template v-if="session.browserWithLedgerSupport">
      <div class="session-main">
        <HardwareState :loading="status === `connect` ? false : true">
          <template v-if="status === `connect` || status === `detect`">
            Please plug in your Ledger&nbsp;Nano&nbsp;S and open the Cosmos app
          </template>
          <p v-if="connectionError" class="error-message">
            {{ connectionError }}
          </p>
        </HardwareState>
      </div>
      <div class="session-footer">
        <p class="ledger-install">
          If you don't have a Ledger Nano, you can
          <a href="" target="_blank" rel="noopener norefferer">buy one here</a>.
        </p>
        <TmBtn
          :value="submitCaption"
          :disabled="status === `connect` ? false : `disabled`"
          @click.native="signIn()"
        />
      </div>
    </template>

    <div v-else class="session-main">
      <p>
        Please use Chrome, Opera, or Brave. Ledger is not supported in this
        browser.
      </p>
    </div>
  </div>
</template>

<script>
import TmBtn from "common/TmBtn"
import { mapGetters } from "vuex"
import HardwareState from "common/TmHardwareState"
export default {
  name: `session-hardware`,
  components: {
    TmBtn,
    HardwareState
  },
  data: () => ({
    status: `connect`,
    connectionError: null,
    address: null
  }),
  computed: {
    ...mapGetters([`session`]),
    submitCaption() {
      return {
        connect: "Sign In",
        detect: "Waiting for Ledger"
      }[this.status]
    }
  },
  methods: {
    setState(value) {
      this.$emit(`route-change`, value)
    },
    goBack() {
      this.$emit(`route-change`, "existing")
    },
    close() {
      this.$emit(`close`)
    },
    async signIn() {
      this.connectionError = null
      this.status = `detect`
      this.address = null
      try {
        this.address = await this.$store.dispatch(`connectLedgerApp`)
      } catch ({ message }) {
        this.status = `connect`
        this.connectionError = message
        return
      }

      await this.$store.dispatch(`signIn`, {
        sessionType: `ledger`,
        address: this.address
      })
    }
  }
}
</script>
<style scoped>
.error-message {
  color: var(--danger);
  font-size: var(--sm);
  font-style: italic;
  margin-bottom: 0;
  padding-top: 1rem;
}

.install-notes {
  flex-direction: column;
}

.ledger-install {
  font-size: var(--sm);
  margin-bottom: 0;
}

.session-footer {
  justify-content: space-between;
}

.address {
  color: var(--link);
  font-weight: 500;
  font-size: 14px;
  white-space: nowrap;
}
</style>
