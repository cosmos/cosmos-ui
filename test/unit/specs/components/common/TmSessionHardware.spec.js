import Vuex from "vuex"
import Vuelidate from "vuelidate"
import { mount, createLocalVue } from "@vue/test-utils"
import TmSessionHardware from "common/TmSessionHardware"

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(Vuelidate)
localVue.directive(`tooltip`, () => {})
localVue.directive(`focus`, () => {})

describe(`TmSessionHardware`, () => {
  let wrapper, store

  beforeEach(() => {
    store = new Vuex.Store()
    wrapper = mount(TmSessionHardware, {
      localVue,
      store
    })
    store.commit = jest.fn()
  })

  describe(`has the expected html structure`, () => {
    it(`with no errors`, () => {
      expect(wrapper.vm.$el).toMatchSnapshot()
    })

    it(`when there're errors`, async () => {
      wrapper.vm.connectionError = `No Ledger found`
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.$el).toMatchSnapshot()
    })
  })

  it(`should go back to the welcome screen on click`, () => {
    wrapper
      .findAll(`.tm-session-header a`)
      .at(0)
      .trigger(`click`)
    expect(store.commit.mock.calls[0][0]).toBe(`setSessionModalView`)
    expect(store.commit.mock.calls[0][1]).toBe(`welcome`)
  })

  it(`should close the session modal`, () => {
    wrapper
      .findAll(`.tm-session-header a`)
      .at(1)
      .trigger(`click`)
    expect(store.commit.mock.calls[0][0]).toBe(`toggleSessionModal`)
    expect(store.commit.mock.calls[0][1]).toBe(false)
  })

  it(`sets the step status`, () => {
    const self = { step: `connect` }
    TmSessionHardware.methods.setStatus.call(self, `detect`)
    expect(self.status).toBe(`detect`)
  })

  it(`sets a connection error`, () => {
    const self = { connectionError: null }
    TmSessionHardware.methods.setConnectionError.call(self, `No Ledger found`)
    expect(self.connectionError).toBe(`No Ledger found`)
  })

  it(`should show a state indicator for different states of the hardware connection`, () => {
    wrapper.setData({ status: `connect` })
    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setData({ status: `detect` })
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe(`tries connecting to Ledger`, () => {
    it(`connects if Ledger is connected and app is open `, async () => {
      const $store = { commit: jest.fn(), dispatch: jest.fn() }
      const self = {
        $store,
        status: `connect`,
        connectionError: null,
        setStatus: jest.fn(),
        setConnectionError: jest.fn(error => (self.connectionError = error))
      }
      await TmSessionHardware.methods.connectLedger.call(self)
      expect(self.$store.dispatch).toHaveBeenCalledWith(`connectLedgerApp`)
      expect(self.connectionError).toBeNull()
    })

    it(`doesn't connect otherwise`, async () => {
      const $store = {
        commit: jest.fn(),
        dispatch: jest.fn(async () =>
          Promise.reject(new Error(`No Ledger found`))
        )
      }
      const self = {
        $store,
        status: `connect`,
        connectionError: null,
        setStatus: jest.fn(),
        setConnectionError: jest.fn(error => (self.connectionError = error))
      }
      await TmSessionHardware.methods.connectLedger.call(self)
      expect(self.$store.dispatch).toHaveBeenCalledWith(`connectLedgerApp`)
      expect(self.connectionError).toBe(`No Ledger found`)
    })
  })
})
