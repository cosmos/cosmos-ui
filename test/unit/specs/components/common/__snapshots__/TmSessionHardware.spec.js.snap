// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmSessionHardware has the expected html structure when there're errors 1`] = `
<div
  class="tm-session"
>
  <div
    class="tm-session-container"
  >
    <div
      class="tm-session-header"
    >
      <a>
        <i
          class="material-icons"
        >
          arrow_back
        </i>
      </a>
       
      <div
        class="tm-session-title"
      >
        
        Sign In
      
      </div>
       
      <a>
        <i
          class="material-icons"
        >
          help_outline
        </i>
      </a>
    </div>
     
    <div
      class="tm-session-main"
    >
      <div
        class="tm-hardware-state"
      >
        <!---->
         
        <i
          class="tm-hardware-state__icon material-icons"
        >
          usb
        </i>
         
        <div
          class="tm-hardware-state__label"
        >
          Please plug in your Ledger Nano S and open the Cosmos app
        </div>
      </div>
       
      <!---->
    </div>
     
    <div
      class="tm-session-footer"
    >
      <p
        class="tm-form-msg sm tm-form-msg--error"
      >
        
        No Ledger found
      
      </p>
    </div>
  </div>
</div>
`;

exports[`TmSessionHardware has the expected html structure with no errors 1`] = `
<div
  class="tm-session"
>
  <div
    class="tm-session-container"
  >
    <div
      class="tm-session-header"
    >
      <a>
        <i
          class="material-icons"
        >
          arrow_back
        </i>
      </a>
       
      <div
        class="tm-session-title"
      >
        
        Sign In
      
      </div>
       
      <a>
        <i
          class="material-icons"
        >
          help_outline
        </i>
      </a>
    </div>
     
    <div
      class="tm-session-main"
    >
      <div
        class="tm-hardware-state"
      >
        <!---->
         
        <i
          class="tm-hardware-state__icon material-icons"
        >
          usb
        </i>
         
        <div
          class="tm-hardware-state__label"
        >
          Please plug in your Ledger Nano S and open the Cosmos app
        </div>
      </div>
       
      <!---->
    </div>
     
    <div
      class="tm-session-footer"
    >
      <!---->
    </div>
  </div>
</div>
`;

exports[`TmSessionHardware should show a state indicator for different states of the hardware connection 1`] = `
"<div class=\\"tm-session\\"><div class=\\"tm-session-container\\"><div class=\\"tm-session-header\\"><a><i class=\\"material-icons\\">arrow_back</i></a> <div class=\\"tm-session-title\\">
        Sign In
      </div> <a><i class=\\"material-icons\\">help_outline</i></a></div> <div class=\\"tm-session-main\\"><div class=\\"tm-hardware-state\\"><!----> <i class=\\"tm-hardware-state__icon material-icons\\">usb</i> <div class=\\"tm-hardware-state__label\\">Please plug in your Ledger Nano S and open the Cosmos app</div></div> <!----></div> <div class=\\"tm-session-footer\\"><!----></div></div></div>"
`;

exports[`TmSessionHardware should show a state indicator for different states of the hardware connection 2`] = `
"<div class=\\"tm-session\\"><div class=\\"tm-session-container\\"><div class=\\"tm-session-header\\"><a><i class=\\"material-icons\\">arrow_back</i></a> <div class=\\"tm-session-title\\">
        Sign In
      </div> <a><i class=\\"material-icons\\">help_outline</i></a></div> <div class=\\"tm-session-main\\"><!----> <div class=\\"tm-hardware-state\\"><img src=\\"~assets/images/loader.svg\\"> <i class=\\"tm-hardware-state__icon material-icons\\"></i> <div class=\\"tm-hardware-state__label\\">Connecting...</div></div></div> <div class=\\"tm-session-footer\\"><!----></div></div></div>"
`;
